js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/countries-price-input-form/index.js

```diff
@@ -16,7 +16,7 @@ const CountriesPriceInputForm = ( props ) => {
 
 	useEffect( () => {
 		setValue( savedValue );
-	}, [ savedValue ] );
+	}, [ savedValue.price ] );
```

Comment:
🚧 React ESLint will complain about this: `"React Hook useEffect has a missing dependency: 'savedValue'. Either include it or remove the dependency array.eslintreact-hooks/exhaustive-deps"`

![image](https://user-images.githubusercontent.com/417342/146390164-eed3a661-284a-47e1-8e66-20c0c5f577b3.png)

Do you have ESLint extension enabled in your editor? Your editor should be able to help you catch this.

Reply:
Sorry, I don’t know why my VS code extension didn’t detect it but it is now working 👍 . I have updated the code as well. Commit: [cf03ab5](https://github.com/woocommerce/google-listings-and-ads/commit/cf03ab591744d709d599fbba77cdf6d9a19b2bd5)
 
I would love to discuss a bit more how the props from the CountriesPriceInputForm component are populated. I think the prop `savedValue` should stay immutable if all the values stay the same. I have been looking deeper into the code and I see that the function getCountriesPriceArray is creating a new array every time, therefore, triggering the useEffect and causing the issue that we are trying to resolve. Please let me know your thoughts about this.

Reply:
> I think the prop `savedValue` should stay immutable if all the values stay the same.

Just a note, there is a hook to help you do that: https://github.com/woocommerce/google-listings-and-ads/blob/ce120cc0704e8f6fca8e9e90c4c54700f26b69d3/js/src/hooks/useIsEqualRefValue.js. I think you have a point there.

> I have been looking deeper into the code and I see that the function getCountriesPriceArray is creating a new array every time, therefore, triggering the useEffect and causing the issue that we are trying to resolve. Please let me know your thoughts about this.

Thanks for this. I looked deeper and found that you are right. Here's my finding below. 

1. The components hierarchy is like this: `ShippingRateSetup` > `CountriesPriceInputForm` > `CountriesPriceInput`.
2. When users change values in `CountriesPriceInput` and then lose focus on the input, `onBlur` event is triggered. 
3. In the `handleBlur` handler in `CountriesPriceInputForm`, first we call `setValue` which causes state change in `CountriesPriceInputForm`, and trigger re-render (1st re-render of `CountriesPriceInputForm`). 
4. After `setValue`, we call `upsertShippingRates` which will call API and then update wp-data redux state. Because of this state change, anything that uses data from `useShippingRates` will also re-render. The parent component `ShippingRateSetup` uses `useShippingRates`, so it and its children `CountriesPriceInputForm` will re-render (2nd re-render of `CountriesPriceInputForm`).
5. In `CountriesPriceInputForm`, because the `savedValue` props is changed, the `useEffect` would be triggered and call `setValue`, causing a state change and re-render again (3rd re-render of `CountriesPriceInputForm`).

Obviously all these isn't ideal. There are probably other issues which I have not discovered or mentioned above.

I think for the long term goal, we may want to address it as a separate PR, probably under issue https://github.com/woocommerce/google-listings-and-ads/issues/288 (under the "Better implementation" section, about putting shipping rates and shipping times values into Form component and pass them down to child components).

Reply:
Thanks, @ecgan  for sharing that hook, I have implemented a new solution using useIsEqualRefValue . Please, let me know what are your thoughts about the new changes: [763cfbc](https://github.com/woocommerce/google-listings-and-ads/commit/763cfbcee56165fb454cbf8e3ab02c654f41d63a)

I see that there was already a discussion about “Setup MC Step 3” implementation and that there is some “future work” pending to be addressed that will help us to improve the design and the performance. Thanks for giving me a bit more context about the project 👍

Reply:
@jorgemd24 , 

> I have implemented a new solution using useIsEqualRefValue . Please, let me know what are your thoughts about the new changes: [763cfbc](https://github.com/woocommerce/google-listings-and-ads/commit/763cfbcee56165fb454cbf8e3ab02c654f41d63a)

I think that looks good. 🙂 👍